scipy/stats/_qmc.py

```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+    the centered discrepancy [1]_. The resulting sample is still a LHS.
+
+    Parameters
+    ----------
+    d : int
+        Dimension of the parameter space.
+    start_sample : array_like (n, d), optional
+        Initial sample to optimize. `LatinHypercube`
+        is used to generate a first design otherwise.
+    n_perturbations : int, optional
+        Number of perturbations to perform. Default is 10000.
+    seed : {None, int, `numpy.random.Generator`}, optional
+        If `seed` is None, the `numpy.random.Generator` singleton is used.
+        If `seed` is an int, a new ``Generator`` instance is used,
+        seeded with `seed`.
+        If `seed` is already a ``Generator`` instance, then that instance is
+        used.
+
+    Notes
+    -----
+    `n_perturbations` set of random permutations are performed aginst a LHS
+    sample. The best design based on the centered discrepancy is constantly
+    update. Centered discrepancy-based design shows better space filling
+    robustness toward 2D and 3D subprojections compared to using other
+    discrepancy measures [2]_.
+
+    References
+    ----------
+    .. [1] Jin et al. "An efficient algorithm for constructing optimal design
+       of computer experiments", Journal of Statistical Planning and
+       Inference, 2005.
+    .. [2] Damblin et al., "Numerical studies of space filling designs:
+       optimization of Latin Hypercube Samples and subprojection properties."
+       Journal of Simulation, 2013.
+
+    Examples
+    --------
+    Generate samples from an optimal design.
+
+    >>> from scipy.stats import qmc
+    >>> sampler = qmc.OptimalLatinHypercube(d=2)
+    >>> sample = sampler.random(n=5)
+    >>> sample
+    array([[0.84052691, 0.53664833],  # random
+           [0.1545328 , 0.36265316],
+           [0.52177809, 0.93343721],
+           [0.68033825, 0.06474907],
+           [0.26544879, 0.61163943]])
+
+    Compute the quality of the sample using the discrepancy criterion.
+
+    >>> qmc.discrepancy(sample)
+    0.01602810655510689  # random
+
+    You can possibly improve the quality of the sample by performing more
+    optimization iterations by using `niter`:
+
+    >>> sampler_2 = qmc.OptimalLatinHypercube(d=2, n_perturbations=5000)
+    >>> sample_2 = sampler_2.random(n=5)
+    >>> qmc.discrepancy(sample_2)
+    0.01600398742881648  # random
+
+    """
+
+    def __init__(self, d: IntNumber, *, start_sample: npt.ArrayLike = None,
+                 n_perturbations: IntNumber = 10000,
+                 seed: SeedType = None):
+        super().__init__(d=d, seed=seed)
+        self.n_perturbations = n_perturbations
+        self.start_sample: Optional[np.ndarray] = None
+
+        if start_sample is not None:
+            self.start_sample = np.asarray(start_sample).copy()
+            self.start_disc = discrepancy(self.start_sample)
+        else:
+            self.start_disc = np.inf
+
+        self.lhs = LatinHypercube(self.d, seed=self.rng)
```

Comment:
This isn't used if there is a `start_sample`, right?

For this and other reasons - e.g. the `n` parameter of `random` has no effect if a `start_sample` are provided - it seems that `start_sample` really throws a wrench into things. I'll reserve judgement for now, but it's hinting that what we really want to have is a function that optimizes a Latin hypercube sample. Maybe it's fine to also have an `OptimalLatinHypercube` class for convenience, but even though intuitively it seems OK for such a class to accept a `start_sample`, it turns out to have some problems.

Reply:
I moved this in `random` as it's only used there and we have no need for it before. When resetting, the same seed will go there, hence the same sampling.

I did not think about `n`... then I agree that this looks a bit weird to have a non-used parameter. So what you are saying is that instead of having this class, I could just refactor `random` to be a standalone function to optimize LHS? This would look the best I guess as it's a one time optimization and we do not need to have an object for that.

Because in the end the convenient class should accept any `QMCEngine` and use `update_discrepancy`. But the interface would be complicated as we would need to allow other cost functions and update functions. We could be re-writting some ersatz of `optimize` 😅 

I will wait for your input here I guess. pinging you here for conveniency @mdhaber

Reply:
Another possibility would be for `LatinHypercube` to return a `LatinHypercubeSample` object, which would essentially just be a `np.ndarray`, but it would have an `optimize` method. (Or it could be an object that holds the array as an attribute, and it has an optimize method). Not sure how SciPythonic this would be, but we are starting to do that sort of thing in `scipy.stats` with result objects that have methods for generating a confidence interval.

Reply:
Nice idea! Do you think that doing so could bring a bit of inconsistency as the other classes return plain `np.ndarray`? Although it could paves the way towards adding new features to the other engines (I still have in mind nice `repr` (optional with plots) and having an object could help, etc.).

What about another method (or `classmethod`) in `LatinHypercube`? This would make it more discoverable.

Reply:
I suppose they can all be `scale`d.

Reply:
All right then I will augment the outputs of the `QMCEngines` and make `random(self, n: int) -> QMCSample` with `QMCSample(np.ndarray)`:
* `discrepancy` property with an optional parameter for the method
* `scale` could scale inplace or not
* (`__repr__` not sure)
* and in this case `optimize`, same can be inplace (should be I guess)

cc @rgommers as even if the submodule is new, this might be surprising for some user and I am not sure if we need/should warn the users before doing this. Or you may have another idea we should explore.

Reply:
I read the code, but kind of missed the actual problem (my brain may not be working anymore). `random()` just returns `best_sample` in all cases it looks like (`start_sample` is also an `(n, d)`-shaped array). Can you spell it out for me?

Reply:
Yes both are `(n,d)`-shaped `np.ndarrays`.

In the end, we can see that `OptimalLatinHypercube` would not really benefit from being a class. Everything is in this `random` function. The feature could either: be a single function; be included in `LatinHypercube` with a new (class?)method; or `random` could return a subclass of `np.ndarray` which will have more functionalities. Doing so, we could harmonise the other engines and we would return a common `QMCSample(np.ndarray)` which would have things like discrepancy and scaling.

Reply:
Ah makes sense. Not a fan of returning an `ndarray` subclass, because subclasses easily get lost due to the `asarray` usage everywhere. The other two options sound better.

Reply:
Ok thanks. I gave this more though and can see these three options to continue (sorry for this, I want to make this right as this is a new module):

1. Add in the class `LatinHypercube` the method `random_optimized(n, *, n_perturbations)`.
I am not sure it's a good idea to allow `n` and `start_sample`. As Matt said, this would make `n` useless if we have `start_sample`. So we might be just using `n` for both. Also maybe keeping this name is better to have the same interface for everyone.

```python
engine = LatinHypercube(d=2)
sample = engine.random_optimized(100)

start_sample = ...
sample = engine.random_optimized(start_sample)
```

2. Just add a parameter `optimize` in `LatinHypercube`.

```python
engine = LatinHypercube(d=2, optimize=True, n_perturbations=10000, start_sample=None)
sample = engine.random(100)

start_sample = ...
engine = LatinHypercube(d=2, optimize=True, n_perturbations=10000, start_sample=start_sample)
sample = engine.random(100)  # but this is a problem as useless param
```

With both of these options, there is the parameter `centered` which would be useless.

3. Going with the function only, advantage: no conflict with `centered` being not used.
```python
sample = latin_hypercube_optimized(d=2, n=100)

sample = latin_hypercube_optimized(d=2, n=start_sample)
```

Reply:
Looking ahead: if there's a change there may be multiple "optimization" strategies, it'd be better as a string rather than a boolean keyword. Not 100% sure that's the case here, but we've made the boolean keyword mistake way too often already.

Reply:
Good point, thanks! Am I reading that you prefer option 2 @rgommers?

Reply:
Good point, thanks! Am I reading that you prefer option 2 @rgommers?

Reply:
Not really, I'm just skimming the discussion - too many other things going on right now. This thing just jumped out at me.

Reply:
@mdhaber do you have any preference? On my side, I think I like 2. And we can also not include `start_sample` for now and see if some people need it.

Reply:
I think the much more common use case will be for someone to want to generate an optimize Latin hypercube - much more often than to optimize an existing Latin hypercube - so I don't think we really need that capability.

With that in mind, adding an argument to the existing method that generates a Latin hypercube (`random`) seems simpler than creating a new function or method. So yes, I think option 2 is the best above.

I agree that we should not add `start_sample` or similar at this time (because of the first point).

I think that `optimize` and `n_perturbations` are redundant for now, if we only have one optimization method. If `n_perturbations` is not 0, then we know the user wants to optimize. We could add an argument to select the optimization method if/when we add another method.

On the other hand, `n_perturbations` might not be an appropriate parameter name for a different optimization method. So an alternative would be to keep only the `optimize` keyword, instead... maybe call it `optimization` so that it's more natural for the value to be the name of the optimization method. In that case, we can use a heuristic to detect convergence rather than having the user specifying the number of permutations. We could add options for control in the future if it is a requested feature and if we have a better idea of what parameters would be common to all optimization methods.

Reply:
Thanks Matt! I will refactor to add the parameter and some stopping logic.

Reply:
Done @mdhaber, let me know what you think 😃